src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -6006,14 +5974,39 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
     }
 
     def typed(tree: Tree, mode: Mode, pt: Type): Tree = {
+      // Extract and insert stabilizing ValDefs (if any) which might have been
+      // introduced during the typing of the original expression.
+      @inline def insertStabilizer(doTyped: Mode => Tree): Tree = {
+        if (phase.erasedTypes || mode.in(APPSELmode) || isMacroImplRef(tree)) doTyped(mode)
+        else tree match {
+          case _: Select | _: Apply | _: TypeApply =>
+            val saved = context.pendingStabilizers
+            context.pendingStabilizers = Nil
+            val (res, vdefs) =
+              try {
+                val r = doTyped(mode | APPSELmode)
+                (r, context.pendingStabilizers)
+              }
+              finally
+                context.pendingStabilizers = saved
+            if (vdefs.isEmpty)
+              res
+            else {
+              devWarningIf(vdefs.forall(_.symbol.owner == context.owner))(s"${context.owner} - ${(vdefs.map(vd => (vd.symbol, vd.symbol.owner.fullNameString)), context.owner)}")
+              typed1(Block(vdefs.reverse, res) setPos res.pos, mode, pt)
```

Comment:
I'm thinking this should be `.setType(res.tpe)` rather than `typed1`. I believe that we either don't need or don't want the extra functionality from `typedBlock`:
  - No need to enter the vdef symbols into the namer, they aren't going to be the target of `typedIdent`.
  - No need to expand macro annotations in the generated stabillizers.
  - The block doesn't match the anonymous class pattern
  - No need to check for pure expressions in statement position

Calling `.setType` would be consistent with named apply blocks: https://github.com/scala/scala/blob/a02eafec13c81c6a21f915306f401ee762af3dc9/src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala#L206-L207

Reply:
That's fine with me. I thought that calling the typer is more regular, and if I remember correctly it helped me shaking out some bugs. I pushed your version now.